### PR TITLE
gh-104469 : Convert _testcapi/dict.c to use AC

### PR DESCRIPTION
**Pull request by jamie2779**
* Issue: gh-104469

### CODE REVIEW

**Comment by jamie2779** [comment]
@corona10

File: Modules/_testcapi/dict.c [hunk 1]
```diff
@@ -165,15 +252,21 @@ dict_getitemref(PyObject *self, PyObject *args)
     }
 }
 
+
+/*[clinic input]
+_testcapi.dict_getitemstringref
+    obj: object
+    value: object
+    attr_name: str(accept={robuffer}, zeroes=True)
```
Comment (corona10) [inline Modules/_testcapi/dict.c]: It looks like `str(accept={str, NoneType}, zeroes=True)` according to https://docs.python.org/3/howto/clinic.html?highlight=clinic#how-to-use-real-argument-clinic-converters-instead-of-legacy-converters

Comment (corona10) [reply Modules/_testcapi/dict.c]: Hi @erlend-aasland and @AlexWaygood  it doesn't work when we use `str(accept={str, NoneType}, zeroes=True)` grammar even if documentation support it. Would you like to take a look?

````
Error in file 'Modules/_testcapi/dict.c' on line 260:
str_converter: illegal combination of arguments (frozenset({<class 'NoneType'>, <class 'str'>}), False, True)
````

File: Modules/_testcapi/dict.c [hunk 2]
```diff
@@ -140,13 +221,19 @@ dict_getitemwitherror(PyObject *self, PyObject *args)
 }
 
 
+/*[clinic input]
+_testcapi.dict_getitemref
+    obj: object
+    attr_name: object
+    value: object
```
Comment (erlend-aasland) [inline Modules/_testcapi/dict.c]: This is not correct; `dict_getitemref` only takes two positional-only arguments: `obj` and `attr_name`.

File: Modules/_testcapi/dict.c [hunk 3]
```diff
@@ -165,15 +252,21 @@ dict_getitemref(PyObject *self, PyObject *args)
     }
 }
 
+
+/*[clinic input]
+_testcapi.dict_getitemstringref
+    obj: object
+    value: object
```
Comment (erlend-aasland) [inline Modules/_testcapi/dict.c]: Ditto. `value` is not part of the spec of this method; `dict_getitemstringref` only takes two positional-only arguments.

File: Modules/_testcapi/dict.c [hunk 4]
```diff
@@ -191,100 +284,159 @@ dict_getitemstringref(PyObject *self, PyObject *args)
...
     NULLABLE(value);
     RETURN_INT(PyDict_SetItem(mapping, key, value));
 }
 
+
+/*[clinic input]
+_testcapi.dict_setitemstring
+    mapping: object
+    value: object
+    key: str(accept={robuffer}, zeroes=True)
```
Comment (erlend-aasland) [inline Modules/_testcapi/dict.c]: The order is incorrect. It should be `mapping, key, value`.

File: Modules/_testcapi/dict.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_testcapi/dict.c]: The docs are inaccurate; the *accept* parameter should be `accept={robuffer, str, NoneType}`. See the definition in clinic.py:

https://github.com/python/cpython/blob/23a6db98f21cba3af69a921f01613bd5f602bf6d/Tools/clinic/clinic.py#L3927

Comment (erlend-aasland) [reply Modules/_testcapi/dict.c]: Also, it is possible to just use the legacy converter:

```suggestion
    attr_name: 'z#'
```

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

**Comment by serhiy-storchaka** [comment]
It significantly reduces readability to me.

It significantly increases the number of lines in `Modules/_testcapi/dict.c`:
```
 Modules/_testcapi/dict.c | 412 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++----------------------------------------------
 1 file changed, 291 insertions(+), 121 deletions(-)
```
And this is not counting the large amount of code added to another file.

I have a limited field of vision and can't see anything beyond a limited number of lines of code. Before that, test functions were compact. After these changes, they will exceed my limit. And if I need to understand what exactly the function does, I have to look for the corresponding code in the generated file and then go back to the previous file, but when I look from the file I lose the context and I have to reread each line on the screen again to find the right place.

These changes are not friendly to the visually impaired.

**Comment by jamie2779** [comment]
@erlend-aasland @corona10 I have made the requested changes; Please Take Another Look!

**Comment by erlend-aasland** [comment]
I think Serhiy's opinion should be taken into account. Converting _testcapi to Argument Clinic is not of importance.

**Comment by ambv** [comment]
Closing and re-opening to retrigger CLA checks. Sorry for the noise.

**Comment by corona10** [comment]
close via https://github.com/python/cpython/issues/104469#issuecomment-1674897104

**Pull request closed without merging**